IncludeToolbox/Formatter/IncludeFormatter.cs

```diff
@@ -178,41 +176,40 @@ private static List<IncludeLineInfo> SortIncludes(IList<IncludeLineInfo> lines,
             else if (typeSorting == FormatterOptionsPage.TypeSorting.QuotedFirst)
                 sortedIncludes = sortedIncludes.OrderBy(x => x.LineDelimiterType == IncludeLineInfo.DelimiterType.Quotes ? 0 : 1);
 
-            // Finally, update the actual lines
+            // Merge sorted includes with original non-include lines
+            var sortedIncludeEnumerator = sortedIncludes.GetEnumerator();
+            var sortedLines = includeBatch.Select(originalLine =>
             {
-                bool firstLine = true;
-
-                foreach (var sortedLine in sortedIncludes)
+                if (originalLine.ContainsActiveInclude)
                 {
-                    // Advance until there is an include line to replace. There *must* be one left if sortedIncludes is not empty.
-                    while (!originalLineEnumerator.Current.ContainsActiveInclude)
-                    {
-                        outSortedList.Add(originalLineEnumerator.Current);
-                        hasElements = originalLineEnumerator.MoveNext();
-                        System.Diagnostics.Debug.Assert(hasElements, "There must be an element left in the original list if there are still sorted elements to put back in.");
-                    }
+                    // Replace original include with sorted includes
+                    return sortedIncludeEnumerator.MoveNext() ? sortedIncludeEnumerator.Current : new IncludeLineInfo();
+                }
+                return originalLine;
+            });
 
-                    bool isLastLine = !originalLineEnumerator.MoveNext();
+            if (settings.RemoveEmptyLines)
+            {
+                // Removing duplicates may have introduced new empty lines
+                sortedLines = sortedLines.Where(sortedLine => !string.IsNullOrWhiteSpace(sortedLine.RawLine));
+            }
 
+            // Finally, update the actual lines
```

Comment:
Instead of doing what we were doing before to insert original non-include lines, this merges those in ahead of time. That way, there's less weird broken logic trying to do it on the fly :)

Reply:
Ohhh I really like the approach with the select that sneaks in the lines from the other enumerator. Never occurred to me to be honest